### PR TITLE
Feature/add date to habit

### PR DESCRIPTION
**Pull request by knite51**
#### What does this PR do?
- add additional columns (Expiry Date and Number of Days) to the Habit table.

#### Description of Tasks to be completed?
- create migration file to add new columns to the habits table in the database
-  refactor the Habit Controller and add new column properties
- validate new input parameters
- write the necessary test for this effect

#### What are the relevant pivotal tracker stories?
- https://www.pivotaltracker.com/story/show/163029971

#### Blockers (if any)

### CODE REVIEW

File: models/habits.js [hunk 1]
```diff
@@ -5,27 +5,35 @@ const HabitsModel = (sequelize, DataTypes) => {
...
       type: DataTypes.STRING,
       allowNull: false
     },
     userId: {
-      type: DataTypes.UUID,
+      type: DataTypes.UUID
+    },
+    expiresAt: {
+      type: DataTypes.STRING,
+      allowNull: true
```
Comment (ghost) [inline models/habits.js]: Just to clarify, this suggests that it is optional to provide an expiry date, yes?

Comment (knite51) [reply models/habits.js]: Oh, it shouldn't be, I will change it false.

File: routes/habit.js [hunk 2]
```diff
@@ -11,8 +11,13 @@ const router = express.Router();
 const { authenticateUser } = authentication;
 const { authorizeHabitOwnerOrAdmin, authorizeHabitOwner } = authorization;
 const {
+  ensureParamsAreNotEmpty,
   ensureNameIsProvided,
   ensureNameIsNotEmpty,
+  ensureNumberDaysAreProvided,
+  ensureNumberDaysNotEmpty,
+  ensureExpiryDateProvided,
```
Comment (ghost) [inline routes/habit.js:19]: This answers my earlier question. So it means that `allowNull` in the models should be false then.

**Comment by ghost** [comment]
Great work altogether.

An observation/suggestion, however, is that I feel that `startDate` may be a useful column to be included in the habits table, and the reasoning is this:

There should be the option to decide which day to kickstart the milestones for an habit even when you have entered it. Does not necessarily have to default to the current date. The kick is that if we have a defined start date (with the default set to be the day on which the habit was created if not edited), then even the `daysBeforeExpiration` may be automatically calculated as each day passes on the frontend with the aid of `moment` or some other normal arithmetic without having to get that data from the DB.

Just my thoughts though.

**Comment by knite51** [comment]
Yes, thank you for the suggestion about the startDate. Skipped my thoughts totally. I will put that in place.

**Comment by ghost** [comment]
@knite51 I noticed that the migration file you just created isn't exactly positioned as the last/latest migration file. Could you update the name so that it follows accordingly please?

**Comment by knite51** [comment]
> @knite51 I noticed that the migration file you just created isn't exactly positioned as the last/latest migration file. Could you update the name so that it follows accordingly, please?

oh okay. I actually wanted to ask how you named the milestone migration file. I believe its (year, month, day, hr, min, sec) so the extra digit 38 is gotten from where exactly?

**Comment by ghost** [comment]
Oh yeah, Ideally supposed to be `YYYYMMDDHHMMSS` There must have been a mistake with the extra `8` at the end of the `create-milestone`

However, there is no real convention that says it must be date. What matters is that alphabetically or numerically, the latest one always comes last. So if for example, I create a new migration and I simply change the 8 in `create-milestone` to `9` in mine, then I am in order.

**Comment by knite51** [comment]
Okay no whala

**Review by ghost — APPROVED** [review]

**Pull request merged**